### PR TITLE
jsgen improvements for ES module support, JS decorators etc

### PR DESCRIPTION
**Pull request by kristianmandrup**
Work In Progress - PR open for discussion, suggestions etc (playing with experimental jsgen infrastructure changes)

This PR now has a Nim [RFC](https://github.com/nim-lang/RFCs/pull/195)

Allows use of ES imports/exports

```nim
import macros, jsffi
{.emit: "/*IMPORTSECTION*/import { x  } from './x'".}

var
  nimx: seq[int]
{.emit: "%ID% = x".}
```

Outputs

```js
import { x  } from './x'
var nim_program_result = 0;
// ...
var nimx_128014 = [null];
nimx_128014 = x
```

Where `x_128014` in JS land represents the `nimx` var in Nim land, assigned to the imported `x` after declaration. We can use the `%ID%` reference for similar effect to achieve clean ES module exports.

Features:

- Add ability to set extension of output file(s) via `js:[ext]` argument to nim compiler
- Output additional supporting files as side effects
- Emit to specific sections of a given output file
- reference to latest declaration output name via `%ID%`

## ID ref in pragma emit

Allows use of `%ID%` to reference latest output declaration name

## JS mode

Allows control of specific output file extension (`ts`, `js`, `tsx`, `jsx`, ...)

`$ nim js abc.nim` -> `abc.js`
`$ nim js abc.nim --jsmode mjs` -> `abc.mjs` (NodeJS with ES modules)
`$ nim js abc.nim --jsmode ts` -> `abc.ts` (TypeScript)

## JS supporting output files

The emit pragma allows a prefix of `/*FILEPATH:[file name]:*/` such as `/*FILEPATH:index.d.ts:*/` to tell the emit which supporting file to emit to.

Enables outputting multiple supporting files, not just a single js file (useful in contexts such as TypeScript with a `d.ts` being built up that contains type definitions/interfaces for better TypeScript interop

## JS sections

Adds Javascript sections: `header`, `import`, `types`, `footer`, similar to concept used in `cgen`

- output to `header` section when `/*HEADERSECTION*/` emit prefix
- output to `import` section when `/*IMPORTSECTION*/` emit prefix
- output to `footer` section when `/*FOOTERSECTION*/` emit prefix
- output to `types` section when `/*TYPESECTION*/` emit prefix

Reasoning

- section `header` can be used for header comment (such as copyright)
- section `import` can be used for ES module imports (and `require` etc for other module formats)
- section `footer` useful for exports
- section `types` useful for typescript

The order will by default be:

- `getHeader()` - compiler internal 
- `header` - author comment header
- `import` - ES module imports
- `types` - for improved typescript interop (`--jsmode ts`)
- `code`
- `footer`

The extra sections are available as pragma emit targets to ensure specific order when required by JS compiler

## Inject before or after var declaration

Add ability to inject code before or after a var declaration

## Structural JS generator improvements

`code`, `header`, `imports`, `footer` and `types` are all of `PSrcCode` instead of simple `Rope`

The `PSrcCode` contains a sequence of Rope that are concatenated on final output write.
`PSrcCode` also has a number of convenient utility methods for more flexibility and power when building the output src code.

```nim
  TSrcCode = object
    srcList: seq[Rope] 

  PGlobals = ref object of RootObj
    lastVarName: Rope
    code, header, imports, footer, types: PSrcCode
```

## Declaration lookup table and before/after injection

Lookup table for maintaining pointers to where declarations have been made in the SrcCode lists (ie index in `srcList: seq[Rope]`.

```nim
  TIdTableEntry = object
    startIndex: int
    endIndex: int

  PIdTableEntry = ref TIdTableEntry

  TLookupTable = object
    idMap: TableRef[string, PIdTableEntry]
    currentId: string
    current: PIdTableEntry

  PLookupTable = ref TLookupTable
```

Currently the `idLookupTable` is set on the `TProc` object for the `p: PProc`, passed around as the compiler context.

```nim
  TProc = object
    procDef: PNode
    idLookupTable: PLookupTable
```

The driver for this feature is to enable better interop with modern Ecmascript libraries, where a LOT of libraries already take advantage of experimental JS decorators.

## Generic ID Ref

Instead of relying on a "hard-coded" `lastVarName` that can be referenced with `%ID%` it would be better with a more flexible approach that could be used with various declaration types, such as: `class`, `method`, `property`, `function`, `variable`, ...

We could enable another emit of the form `{.emit: "/*STOREID:class*/}` to store the latest declaration metadata in an id type table with a particular key, such as `class` in this case.

We could then have a id type table, where each entry is a lookup table of ids for that type of identifier.

```nim
  TIdTableEntry = object
    nimId: string # use s.name.s
    startIndex: int
    endIndex: int

  PTIdTableEntry = ref TIdTableEntry

  TLookupTable = object
    idMap: TableRef[string, PIdTableEntry]

  PLookupTable = ref TLookupTable

  TLookupTypeTable = object
    idTypeMap: TableRef[string, PLookupTable]
```

To get hold of the Nim variable declaration name:

```nim
varName = mangleName(p.module, v)
let nimVarName = v.name.s # to store the nim var name (useful for clean module exports)
```

We could then retrieve the particular value using another emit of the form `%GETID:class%` or even `%GETID:class(x)%` where `x` is the nim declaration id.

### CODE REVIEW

**Comment by timotheecour** [comment]
why not simply reuse -o instead of new syntax `js:mjs` ?
```nim
nim js:mjs abc.nim
=>
nim js -o:abc.mjs abc.nim
```

**Comment by kristianmandrup** [comment]
@timotheecour Yes, that works as well. However I find it slightly more convenient (and safe) to have a way to simply set the output extension, without having to know all the options and write the entire filepath for the `output` option. 

This extension could also be used to internally set the default "target environment", such as when multiple files are output as a result (new emit feature I'm experimenting with).

**Comment by timotheecour** [comment]
> This extension could also be used to internally set the default "target environment",

you can use something like (in pseudocode):
```nim
if mode == "js":
  let ext = outFile.splitFile.ext
  if ext == ".mjs": ...
```

or, you can introduce a new flag: `nim js --jsmode:mjs`
but `js:mjsj` is not as good IMO

**Comment by kristianmandrup** [comment]
Yes, I agree. I would also have preferred an extra option. The tweak I made was just the easiest way I could find and then I thought it was not as bas, but definitely a little dirty. 

Could not figure out how to add an extra option. The compiler code with all the different config options is difficult to track and grasp. I also like your mode solution. 

Can you point me in the right direction? :)

**Comment by timotheecour** [comment]
* step 1: search where commands are processed; eg: search for "fullhelp" in source code (or any other nim command line flag): this points to compiler/commands.nim as only possibility
* step 2: now you have a a bunch of commands and logic to handle these; find some cmdline flag that behaves similarly to jsmode (ie, that accepts a string, unlike, say, `errormax` which accepts an int) => lets say this is similar flag is `experimental`
* step 3: adapt code for `--experimental` to `--jsmode`:
`parseEnum[Feature]` => `parseEnum[Jsmode]` with some new Jsmode enum

**Comment by kristianmandrup** [comment]
Would the option be added in `commands.nim`?

```nim
proc testCompileOptionArg*(conf: ConfigRef; switch, arg: string, info: TLineInfo): bool =
  case switch.normalize
  of "gc":
    case arg.normalize
    of "boehm": result = conf.selectedGC == gcBoehm
    of "refc": result = conf.selectedGC == gcRefc
    of "v2": result = false
    # ...
  of "opt":
    case arg.normalize
    of "speed": result = contains(conf.options, optOptimizeSpeed)

```

Used from `semfold` but that seems to be loading of some other options, perhaps from `nim.cfg`?

```nim
  of mCompileOption:
    result = newIntNodeT(toInt128(ord(commands.testCompileOption(g.config, a.getStr, n.info))), n, g)
  of mCompileOptionArg:
    result = newIntNodeT(toInt128(ord(
      testCompileOptionArg(g.config, getStr(a), getStr(b), n.info))), n, g)
```

Then there is this proc, also in `commands.nim`

```nim
proc testCompileOption*(conf: ConfigRef; switch: string, info: TLineInfo): bool =
  case switch.normalize
  of "debuginfo": result = contains(conf.globalOptions, optCDebug)
  of "compileonly", "c": result = contains(conf.globalOptions, optCompileOnly)
```

Looks more "legit". But what happens here? does contains test for the presence of the flag, then add the flag to result?

**Comment by timotheecour** [comment]
* `testCompileOption` is (IIRC) for `compileOption("assertions")` introspection => feel free to add a comment (maybe in another PR) to clarify this for that proc; also, it's out of sync and misses many options
* `testCompileOptionArg` is also for introspection, but for for cmdline flags that take a non-bool argument (same comment as above)
* the answer is none of these 2: it's `processSwitch` (which is also where "fullhelp" is processed, as i mentioned above)


when in doubt, use printf/echo-like debugging (or use lldb/gdb); maybe even `getStacktrace` (after compiling nim with --stacktrace:on --debugger:native) so you know how you end up in a given location

**Comment by kristianmandrup** [comment]
Oh, well a bit of a jungle to me. Would be nice with some Nim Docs on at least some of the key procs. Thanks

**Comment by timotheecour** [comment]
well now that I've added `compiler/**` to docgen there's no excuse not to improve doc comments for compiler sources ;-) => https://nim-lang.github.io/Nim/compiler/commands.html is the docgen'd compiler/commands.nim

these are also relevant:
* https://nim-lang.github.io/Nim/nimc.html
* https://nim-lang.github.io/Nim/backends.html

**Comment by kristianmandrup** [comment]
I'm adding `jsmode` as a switch as per your instructions:

```nim
proc processSwitch*(switch, arg: string, pass: TCmdLinePass, info: TLineInfo;
                    conf: ConfigRef) =
  var
    key, val: string
  case switch.normalize
  of "path", "p":
    # ...
  of "jsmode":
    expectArg(conf, switch, arg, pass, info)
    conf.jsMode = if arg.len == 0: "js" else: arg
```

**Comment by kristianmandrup** [comment]
Branch compiles now, but I get this error:

```sh
Nim git:(js-gen-improvements) ./build_all.sh

Hint: 152385 LOC; 11.823 sec; 511.156MiB peakmem; Release build; proj: /Users/kristianmandrup/repos/nim-projects/Nim/compiler/nim.nim; out: /Users/kristianmandrup/repos/nim-projects/Nim/compiler/nim [SuccessX]
compiler/nim1 jsonscript -d:release --nimcache:nimcache/r_macosx_amd64 compiler/nim.nim
Hint: used config file '/Users/kristianmandrup/repos/nim-projects/Nim/config/nim.cfg' [Conf]
Hint: used config file '/Users/kristianmandrup/repos/nim-projects/Nim/compiler/nim.cfg' [Conf]
Hint: used config file '/Users/kristianmandrup/repos/nim-projects/Nim/config/config.nims' [Conf]
fatal.nim(55)            sysFatal
Error: unhandled exception: index 1 not in 0 .. 0 [IndexError]
FAILURE
```

**Comment by kristianmandrup** [comment]
This PR compiles now. Time to test the functionality and add tests for it :)

**Comment by kristianmandrup** [comment]
Will debug and fix shortly...

```sh
$ /home/build/Nim/bin/nim js --hints:on -d:testing --nimblePath:tests/deps --nimCache:nimcache/tests/arithm/tashr.nim_5bc06f5800d415cc95e1349edbaca425 tests/arithm/tashr.nim
Hint: used config file '/usr/home/build/Nim/config/nim.cfg' [Conf]
Hint: used config file '/usr/home/build/Nim/config/config.nims' [Conf]
Hint: used config file '/usr/home/build/Nim/tests/config.nims' [Conf]
Hint: system [Processing]
jsGenPragma
jsGenPragma
jsGenPragma
jsGenPragma
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

**Comment by timotheecour** [comment]
@kristianmandrup just my 2 cents; since this is your 2nd nim PR (other 1 not merged yet), and since this PR is rather complex, before you devote too much time on this, you may want to check with @araq whether this has a chance of being eventually accepted; alternatively you could also tackle getting smaller / less complex improvements merged in nim first before doing heavier work like this one.

as for 
> SIGSEGV: Illegal storage access. (Attempt to read from nil?)

simply compile nim with --stacktrace:on --debugger:native and it should help find out where is SIGSEGV

**Comment by kristianmandrup** [comment]
Of course. For now just focusing on making it work. Then will someone smaller, clean PRs with my suggestions.

It compiles just fine now and main functionality I was striving for works.

File: testfile.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/* Generated by the Nim Compiler v1.1.1 */
```
Comment (timotheecour) [inline testfile.js]: we shouldn't checkin generated files (unless in rare cases such as regression tests)

File: testament/testament.nim [hunk 2]
```diff
@@ -727,7 +744,8 @@ proc main() =
     isMainProcess = false
     var cat = Category(p.key)
     p.next
-    processCategory(r, cat, p.cmdLineRest.string, testsDir, runJoinableTests = false)
+    processCategory(r, cat, p.cmdLineRest.string, testsDir,
```
Comment (timotheecour) [inline testament/testament.nim:747]: reformatting in unrelated parts of code should not be part of this PR but be done in separate PR (and only if it's via running nimpretty with default options); ditto wherever relevant.
The reason is this makes it hard to see actual vs pure reformatting changes

File: compiler/trees.nim [hunk 3]
```diff
@@ -124,8 +124,16 @@ proc isRange*(n: PNode): bool {.inline.} =
       result = true
 
 proc whichPragma*(n: PNode): TSpecialWord =
+  # echo "whichPragma"
```
Comment (timotheecour) [inline compiler/trees.nim]: don't keep temporary debugging statements (or mark the PR as WIP and add something searchable to make it clear to reviewer to ignore those as well as for you to remember to remove in final version, eg:
`# echo "key"`
=>
`# echo "key" # PRTEMP`

File: compiler/main.nim [hunk 4]
```diff
@@ -204,11 +211,12 @@ proc mainCommand*(graph: ModuleGraph) =
       commandCompileToC(graph)
     else:
       rawMessage(conf, errGenerated, "'run' command not available; rebuild with -d:tinyc")
-  of "js", "compiletojs":
```
Comment (timotheecour) [inline compiler/main.nim:207]: why was `compiletojs` removed? note that we also have: `compiletoc` etc

**Comment by kristianmandrup** [comment]
This PR will be fully discarded and redone in a set of small PRs later this week. This PR is mainly to showcase some ideas.

File: compiler/commands.nim [hunk 5]
```diff
@@ -407,6 +407,9 @@ proc processSwitch*(switch, arg: string, pass: TCmdLinePass, info: TLineInfo;
     let f = splitFile(processPath(conf, arg, info, notRelativeToProj=true).string)
     conf.outFile = RelativeFile f.name & f.ext
     conf.outDir = toAbsoluteDir f.dir
+  of "jsmode":
```
Comment (timotheecour) [inline compiler/commands.nim:410]: this should be documented in --help (see `advopt.txt`)

**Comment by timotheecour** [comment]
left some review comments but again, before spending time fixing them, make sure this will eventually get greenlighted

File: compiler/main.nim (hunk 4 above)
Comment (kristianmandrup) [reply compiler/main.nim:207]: Oops, introducing it again

Comment (timotheecour) [reply compiler/main.nim:207]: feel free to click "resolved" when you've addressed a comment (unless for cases where the way you've resolved it could potentially be controversial; obviously not the case here)

**Comment by Araq** [comment]
I'm sure what you do is important work but it would be so much better if you could please write an RFC here https://github.com/nim-lang/RFCs. (https://github.com/nim-lang/Nim/pull/13381#issue-373434878 is a good start for it)

File: testament/testament.nim (hunk 2 above)
Comment (kristianmandrup) [reply testament/testament.nim:747]: I will rework this PR as a set of smaller PRs and make sure to disable the IDE/editor prettifier to avoid auto reformatting.

File: testfile.js (hunk 1 above)
Comment (kristianmandrup) [reply testfile.js]: I've removed it

**Comment by kristianmandrup** [comment]
Will do. Been mostly experimenting and getting to know the internals of the Nim compiler and learning Nim in the process.

**Comment by juancarlospaco** [comment]
This is nice, and the JS target needs more love, hows the development going @kristianmandrup ?.
:)

**Comment by g5becks** [comment]
I really wish I knew more about compilers, but imo - this is the "killer feature" that nim actually needs.

@kristianmandrup  hope this is still in the pipeline! 

I don't think typescript output is a big deal - a  d.ts file is.  To be honest, I'd settle for es6 module outputs and imports, along with typescript declaration file outputs, the rest of the features are just icing.

**Comment by kristianmandrup** [comment]
At the time there was little interest from the core team in improving the nim javascript compilation, a very secondary concern. So I "gave up" on taking this any further. 

The original JS compiler uses simple string concatenation, which doesn't scale in terms of adding anything beyond the most simple functionality. 

In this branch I re-engineered the compiler to build up an array of objects, with index maps, kind of like a very primitive single table Relational DB. The compiler then traverses this built up array at the end, converts each object to a resulting string which becomes the compile output. Depending on the desired target output is can also include some conditional logic.

I implemented all the features listed above successfully with no knowledge of the compiler, using trial and error and a little assistance by some from the core team. 

If anyone is interested in these features, have a look and try to port the parts of the functionality you would like. The underlying mechanics of the compiler needs to be improved, similar to what I attempted in order for the JS compiler to "take off" IMO. Needs some serious work, love and attention. Would love to see a mature JS compiler by version 1.3-2.0.

Cheers ;)

**Comment by Araq** [comment]
> At the time there was little interest from the core team in improving the nim javascript compilation, a very secondary concern. So I "gave up" on taking this any further.

There is interest, however "I rewrote the JS codegen, tests are still green, but nobody knows what it really breaks and we don't know if the author of the new JS backend has the time and will to fix the potential regressions" isn't particularly exciting. There are some ways we can bring this forward: 

- Offer a switch to use the old JS codegen for a migration period.
- Make a ``js2`` command so that people can test their code via ``nim js2 x.nim``
- Use the compiler frontend as a library and provide a ``nimjs`` compiler. Can live in Nim's repository or in a new one.

**Comment by timotheecour** [comment]
also I'll re-iterate my above comment https://github.com/nim-lang/Nim/pull/13381#issuecomment-590141387

> just my 2 cents; since this is your 2nd nim PR (other 1 not merged yet), and since this PR is rather complex, before you devote too much time on this, you may want to check with @Araq whether this has a chance of being eventually accepted; alternatively you could also tackle getting smaller / less complex improvements merged in nim first before doing heavier work like this one.
